### ISSUE TITLE
gcc error with cpython

### ISSUE THREAD

**Issue by Huan-Yang** [bmcage/odes#70]
Hi, I am trying to use Performance tests.ipynb with Cython to check the speedup.
However, I encountered CompileError, I think that is due to Cython/gcc.
Any idea about this? Thanks.

---------------------------------------------------------------------------
```
CompileError                              Traceback (most recent call last)
<ipython-input-15-12eebe5eed09> in <module>()
----> 1 get_ipython().run_cell_magic(u'cython', u'-I /home/yangh/odes-master/odes/scikits/odes/sundials/ -I /home/yangh/anaconda2/lib/python2.7/site-packages/scikits/odes/sundials/', u"## update include flag -I to point to odes/sundials directory!\nimport numpy as np\nfrom cpython cimport bool\ncimport numpy as np\nfrom scikits.odes.sundials.cvode cimport CV_RhsFunction\n\n#scikits.odes allows cython functions only if derived from correct class\n\ncdef class egfngf_cython_model(CV_RhsFunction):\n    cdef public ts, k, y0, userdata\n    cdef public object name\n    cdef public CV_RhsFunction f_odes\n    cdef public bool has_userdata, has_userdata_odes\n    \n    def __cinit__(self):\n        self.name = 'egfngf_cython'\n        self.ts = np.linspace(0, 120, 121, dtype=float)\n        self.has_userdata = True\n        self.has_userdata_odes = True\n        self.k = np.array([\n                    2.18503E-5,\n                    0.0121008,\n                    1.38209E-7,\n                    0.00723811,\n                    694.731,\n                    6086070.0,\n                    389.428,\n                    2112.66,\n                    1611.97,\n                    896896.0,\n                    32.344,\n                    35954.3,\n                    1509.36,\n                    1432410.0,\n                    0.884096,\n                    62464.6,\n                    185.759,\n                    4768350.0,\n                    125.089,\n                    157948.0,\n                    2.83243,\n                    518753.0,\n                    9.85367,\n                    1007340.0,\n                    8.8912,\n                    3496490.0,\n                    0.0213697,\n                    763523.0,\n                    10.6737,\n                    184912.0,\n                    0.0771067,\n                    272056.0,\n                    0.0566279,\n                    653951.0,\n                    15.1212,\n                    119355.0,\n                    146.912,\n                    12876.2,\n                    1.40145,\n                    10965.6,\n                    27.265,\n                    295990.0,\n                    2.20995,\n                    1025460.0,\n                    0.126329,\n                    1061.71,\n                    441.287,\n                    1.08795E7\n                ], float)\n        \n        self.userdata = self.k\n\n        self.y0 = np.array([\n                    1000,\n                    4560,\n                    80000.0,\n                    0.0,\n                    10000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    600000.0,\n                    0.0,\n                    600000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    120000.0,\n                    0.0,\n                    120000.0,\n                    120000.0,\n                    120000.0,\n                    120000.0\n                ], float)\n\n    cpdef np.ndarray[double, ndim=1] f(self, double t, np.ndarray[double, ndim=1] y, \n                                       np.ndarray[double, ndim=1] k):\n        return np.array([\n            ((-1.0 * k[0] * y[0] * y[2])) + (1.0 * k[1] * y[3]),\n            ((-1.0 * k[2] * y[1] * y[4])) + (1.0 * k[3] * y[5]),\n            ((-1.0 * k[0] * y[0] * y[2])) + (1.0 * k[1] * y[3]),\n            ((1.0 * k[0] * y[0] * y[2]) + (-1.0 * k[1] * y[3])),\n            ((-1.0 * k[2] * y[1] * y[4]) + (1.0 * k[3] * y[5])),\n            ((1.0 * k[2] * y[1] * y[4]) + (-1.0 * k[3] * y[5])),\n            ((-1.0 * k[4] * y[3] * y[6] / (y[6] + k[5])) + (-1.0 * k[6] * y[5] * y[6] / (y[6] + k[7])) + (\n                1.0 * k[8] * y[9] * y[7] / (y[7] + k[9]))),\n            ((1.0 * k[4] * y[3] * y[6] / (y[6] + k[5])) + (1.0 * k[6] * y[5] * y[6] / (y[6] + k[7])) + (\n                -1.0 * k[8] * y[9] * y[7] / (y[7] + k[9]))),\n            ((-1.0 * k[26] * y[19] * y[8] / (y[8] + k[27]))),\n            ((1.0 * k[26] * y[19] * y[8] / (y[8] + k[27]))),\n            ((-1.0 * k[10] * y[7] * y[10] / (y[10] + k[11])) + (1.0 * k[12] * y[28] * y[11] / (y[11] + k[13]))),\n            ((1.0 * k[10] * y[7] * y[10] / (y[10] + k[11])) + (-1.0 * k[12] * y[28] * y[11] / (y[11] + k[13]))),\n            ((-1.0 * k[14] * y[11] * y[12] / (y[12] + k[15])) + (1.0 * k[44] * y[31] * y[13] / (y[13] + k[45])) + (\n                1.0 * k[34] * y[23] * y[13] / (y[13] + k[35]))),\n            ((1.0 * k[14] * y[11] * y[12] / (y[12] + k[15])) + (-1.0 * k[44] * y[31] * y[13] / (y[13] + k[45])) + (\n                -1.0 * k[34] * y[23] * y[13] / (y[13] + k[35]))),\n            ((-1.0 * k[42] * y[27] * y[14] / (y[14] + k[43])) + (1.0 * k[46] * y[31] * y[15] / (y[15] + k[47]))),\n            ((1.0 * k[42] * y[27] * y[14] / (y[14] + k[43])) + (-1.0 * k[46] * y[31] * y[15] / (y[15] + k[47]))),\n            ((-1.0 * k[16] * y[13] * y[16] / (y[16] + k[17])) + (-1.0 * k[18] * y[15] * y[16] / (y[16] + k[19])) + (\n                1.0 * k[20] * y[30] * y[17] / (y[17] + k[21]))),\n            ((1.0 * k[16] * y[13] * y[16] / (y[16] + k[17])) + (1.0 * k[18] * y[15] * y[16] / (y[16] + k[19])) + (\n                -1.0 * k[20] * y[30] * y[17] / (y[17] + k[21]))),\n            ((-1.0 * k[22] * y[17] * y[18] / (y[18] + k[23])) + (1.0 * k[24] * y[30] * y[19] / (y[19] + k[25]))),\n            ((1.0 * k[22] * y[17] * y[18] / (y[18] + k[23])) + (-1.0 * k[24] * y[30] * y[19] / (y[19] + k[25]))),\n            ((-1.0 * k[28] * y[3] * y[20] / (y[20] + k[29])) + (-1.0 * k[30] * y[11] * y[20] / (y[20] + k[31]))),\n            ((1.0 * k[28] * y[3] * y[20] / (y[20] + k[29])) + (1.0 * k[30] * y[11] * y[20] / (y[20] + k[31]))),\n            ((-1.0 * k[32] * y[21] * y[22] / (y[22] + k[33]))),\n            ((1.0 * k[32] * y[21] * y[22] / (y[22] + k[33]))),\n            ((-1.0 * k[36] * y[5] * y[24] / (y[24] + k[37]))),\n            ((1.0 * k[36] * y[5] * y[24] / (y[24] + k[37]))),\n            ((-1.0 * k[38] * y[25] * y[26] / (y[26] + k[39])) + (1.0 * k[40] * y[29] * y[27] / (y[27] + k[41]))),\n            ((1.0 * k[38] * y[25] * y[26] / (y[26] + k[39])) + (-1.0 * k[40] * y[29] * y[27] / (y[27] + k[41]))),\n            0,\n            0,\n            0,\n            0], float)\n\n    \n    cpdef int evaluate(self, double t,\n                       np.ndarray[double, ndim=1] y,\n                       np.ndarray[double, ndim=1] yout,\n                       object userdata = None) except? -1:\n        #cdef np.ndarray[double, ndim=1] k = self.k  # avoid self.k gives quite some speedup!\n        cdef np.ndarray[double, ndim=1] k = userdata\n        # avoiding creation of temporary arrays gives quite some speedup!\n        yout[0] = ((-1.0 * k[0] * y[0] * y[2])) + (1.0 * k[1] * y[3])\n        yout[1] = ((-1.0 * k[2] * y[1] * y[4])) + (1.0 * k[3] * y[5])\n        yout[2] = ((-1.0 * k[0] * y[0] * y[2])) + (1.0 * k[1] * y[3])\n        yout[3] = ((1.0 * k[0] * y[0] * y[2]) + (-1.0 * k[1] * y[3]))\n        yout[4] = ((-1.0 * k[2] * y[1] * y[4]) + (1.0 * k[3] * y[5]))\n        yout[5] = ((1.0 * k[2] * y[1] * y[4]) + (-1.0 * k[3] * y[5]))\n        yout[6] = ((-1.0 * k[4] * y[3] * y[6] / (y[6] + k[5])) + (-1.0 * k[6] * y[5] * y[6] / (y[6] + k[7])) + (\n                1.0 * k[8] * y[9] * y[7] / (y[7] + k[9])))\n        yout[7] = ((1.0 * k[4] * y[3] * y[6] / (y[6] + k[5])) + (1.0 * k[6] * y[5] * y[6] / (y[6] + k[7])) + (\n                -1.0 * k[8] * y[9] * y[7] / (y[7] + k[9])))\n        yout[8] = ((-1.0 * k[26] * y[19] * y[8] / (y[8] + k[27])))\n        yout[9] = ((1.0 * k[26] * y[19] * y[8] / (y[8] + k[27])))\n        yout[10] = ((-1.0 * k[10] * y[7] * y[10] / (y[10] + k[11])) + (1.0 * k[12] * y[28] * y[11] / (y[11] + k[13])))\n        yout[11] = ((1.0 * k[10] * y[7] * y[10] / (y[10] + k[11])) + (-1.0 * k[12] * y[28] * y[11] / (y[11] + k[13])))\n        yout[12] = ((-1.0 * k[14] * y[11] * y[12] / (y[12] + k[15])) + (1.0 * k[44] * y[31] * y[13] / (y[13] + k[45])) + (\n                1.0 * k[34] * y[23] * y[13] / (y[13] + k[35])))\n        yout[13] = ((1.0 * k[14] * y[11] * y[12] / (y[12] + k[15])) + (-1.0 * k[44] * y[31] * y[13] / (y[13] + k[45])) + (\n                -1.0 * k[34] * y[23] * y[13] / (y[13] + k[35])))\n        yout[14] = ((-1.0 * k[42] * y[27] * y[14] / (y[14] + k[43])) + (1.0 * k[46] * y[31] * y[15] / (y[15] + k[47])))\n        yout[15] = ((1.0 * k[42] * y[27] * y[14] / (y[14] + k[43])) + (-1.0 * k[46] * y[31] * y[15] / (y[15] + k[47])))\n        yout[16] = ((-1.0 * k[16] * y[13] * y[16] / (y[16] + k[17])) + (-1.0 * k[18] * y[15] * y[16] / (y[16] + k[19])) + (\n                1.0 * k[20] * y[30] * y[17] / (y[17] + k[21])))\n        yout[17] = ((1.0 * k[16] * y[13] * y[16] / (y[16] + k[17])) + (1.0 * k[18] * y[15] * y[16] / (y[16] + k[19])) + (\n                -1.0 * k[20] * y[30] * y[17] / (y[17] + k[21])))\n        yout[18] = ((-1.0 * k[22] * y[17] * y[18] / (y[18] + k[23])) + (1.0 * k[24] * y[30] * y[19] / (y[19] + k[25])))\n        yout[19] = ((1.0 * k[22] * y[17] * y[18] / (y[18] + k[23])) + (-1.0 * k[24] * y[30] * y[19] / (y[19] + k[25])))\n        yout[20] = ((-1.0 * k[28] * y[3] * y[20] / (y[20] + k[29])) + (-1.0 * k[30] * y[11] * y[20] / (y[20] + k[31])))\n        yout[21] = ((1.0 * k[28] * y[3] * y[20] / (y[20] + k[29])) + (1.0 * k[30] * y[11] * y[20] / (y[20] + k[31])))\n        yout[22] = ((-1.0 * k[32] * y[21] * y[22] / (y[22] + k[33])))\n        yout[23] = ((1.0 * k[32] * y[21] * y[22] / (y[22] + k[33])))\n        yout[24] = ((-1.0 * k[36] * y[5] * y[24] / (y[24] + k[37])))\n        yout[25] = ((1.0 * k[36] * y[5] * y[24] / (y[24] + k[37])))\n        yout[26] = ((-1.0 * k[38] * y[25] * y[26] / (y[26] + k[39])) + (1.0 * k[40] * y[29] * y[27] / (y[27] + k[41])))\n        yout[27] = ((1.0 * k[38] * y[25] * y[26] / (y[26] + k[39])) + (-1.0 * k[40] * y[29] * y[27] / (y[27] + k[41])))\n        yout[28] = 0\n        yout[29] = 0\n        yout[30] = 0\n        yout[31] = 0\n        return 0\n    ")

/home/yangh/anaconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc in run_cell_magic(self, magic_name, line, cell)
   2113             magic_arg_s = self.var_expand(line, stack_depth)
   2114             with self.builtin_trap:
-> 2115                 result = fn(magic_arg_s, cell)
   2116             return result
   2117 

<decorator-gen-124> in cython(self, line, cell)

/home/yangh/anaconda2/lib/python2.7/site-packages/IPython/core/magic.pyc in <lambda>(f, *a, **k)
    186     # but it's overkill for just that one bit of state.
    187     def magic_deco(arg):
--> 188         call = lambda f, *a, **k: f(*a, **k)
    189 
    190         if callable(arg):

/home/yangh/anaconda2/lib/python2.7/site-packages/Cython/Build/IpythonMagic.pyc in cython(self, line, cell)
    289             build_extension.build_temp = os.path.dirname(pyx_file)
    290             build_extension.build_lib  = lib_dir
--> 291             build_extension.run()
    292             self._code_cache[key] = module_name
    293 

/home/yangh/anaconda2/lib/python2.7/distutils/command/build_ext.pyc in run(self)
    338 
    339         # Now actually compile and link everything.
--> 340         self.build_extensions()
    341 
    342     def check_extensions_list(self, extensions):

/home/yangh/anaconda2/lib/python2.7/distutils/command/build_ext.pyc in build_extensions(self)
    447 
    448         for ext in self.extensions:
--> 449             self.build_extension(ext)
    450 
    451     def build_extension(self, ext):

/home/yangh/anaconda2/lib/python2.7/distutils/command/build_ext.pyc in build_extension(self, ext)
    497                                          debug=self.debug,
    498                                          extra_postargs=extra_args,
--> 499                                          depends=ext.depends)
    500 
    501         # XXX -- this is a Vile HACK!

/home/yangh/anaconda2/lib/python2.7/distutils/ccompiler.pyc in compile(self, sources, output_dir, macros, include_dirs, debug, extra_preargs, extra_postargs, depends)
    572             except KeyError:
    573                 continue
--> 574             self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
    575 
    576         # Return *all* object filenames, not just the ones we just built.

/home/yangh/anaconda2/lib/python2.7/distutils/unixccompiler.pyc in _compile(self, obj, src, ext, cc_args, extra_postargs, pp_opts)
    122                        extra_postargs)
    123         except DistutilsExecError, msg:
--> 124             raise CompileError, msg
    125 
    126     def create_static_lib(self, objects, output_libname,

CompileError: command 'gcc' failed with exit status 1
```

**Comment by aragilar** [comment]
@Huan-Yang Do you have sundials installed? You need to point the include to wherever you installed sundials.

**Comment by Huan-Yang** [comment]
@aragilar  Thanks for your reply.
In my try, I adapted the %%cython line like
>%%cython -I /home/yangh/odes-master/odes/scikits/odes/sundials/ -I /home/yangh/anaconda2/lib/python2.7/site-packages/scikits/odes/sundials/

The files contained in these two directories are the following. I also wonder whether they are those correct files.
>~/odes-master/odes/scikits/odes/sundials$ ls -al
total 244
drwxrwxr-x 2 yangh yangh  4096 Feb 16 01:27 .
drwxrwxr-x 5 yangh yangh  4096 Feb 16 01:27 ..
-rw-rw-r-- 1 yangh yangh  2169 Feb 16 01:27 __init__.py
-rw-rw-r-- 1 yangh yangh 11520 Feb 16 01:27 c_cvode.pxd
-rw-rw-r-- 1 yangh yangh 11301 Feb 16 01:27 c_ida.pxd
-rw-rw-r-- 1 yangh yangh  2061 Feb 16 01:27 c_nvector_serial.pxd
-rw-rw-r-- 1 yangh yangh 20652 Feb 16 01:27 c_sundials.pxd
-rw-rw-r-- 1 yangh yangh   394 Feb 16 01:27 common_defs.pxd
-rw-rw-r-- 1 yangh yangh  3920 Feb 16 01:27 common_defs.pyx
-rw-rw-r-- 1 yangh yangh  4612 Feb 16 01:27 cvode.pxd
-rw-rw-r-- 1 yangh yangh 75153 Feb 16 01:27 cvode.pyx
-rw-rw-r-- 1 yangh yangh  4047 Feb 16 01:27 ida.pxd
-rw-rw-r-- 1 yangh yangh 71692 Feb 16 01:27 ida.pyx
-rw-rw-r-- 1 yangh yangh  3823 Feb 16 01:27 setup-tine.py
-rw-rw-r-- 1 yangh yangh  4681 Feb 16 01:27 setup.py

>~/anaconda2/lib/python2.7/site-packages/scikits/odes/sundials$ ls -al
total 5352
drwxrwxr-x 2 yangh yangh    4096 Feb 16 01:28 .
drwxrwxr-x 4 yangh yangh    4096 Feb 16 01:28 ..
-rw-rw-r-- 1 yangh yangh    2169 Feb 16 01:27 __init__.py
-rw-rw-r-- 1 yangh yangh    4134 Feb 16 01:28 __init__.pyc
-rw-rw-r-- 1 yangh yangh   11520 Feb 16 01:27 c_cvode.pxd
-rw-rw-r-- 1 yangh yangh   11301 Feb 16 01:27 c_ida.pxd
-rw-rw-r-- 1 yangh yangh    2061 Feb 16 01:27 c_nvector_serial.pxd
-rw-rw-r-- 1 yangh yangh   20652 Feb 16 01:27 c_sundials.pxd
-rw-rw-r-- 1 yangh yangh     394 Feb 16 01:27 common_defs.pxd
-rwxrwxr-x 1 yangh yangh  242024 Feb 16 01:28 common_defs.so
-rw-rw-r-- 1 yangh yangh    4612 Feb 16 01:27 cvode.pxd
-rwxrwxr-x 1 yangh yangh 2623528 Feb 16 01:28 cvode.so
-rw-rw-r-- 1 yangh yangh    4047 Feb 16 01:27 ida.pxd
-rwxrwxr-x 1 yangh yangh 2496288 Feb 16 01:28 ida.so
-rw-rw-r-- 1 yangh yangh    3823 Feb 16 01:27 setup-tine.py
-rw-rw-r-- 1 yangh yangh    3168 Feb 16 01:28 setup-tine.pyc
-rw-rw-r-- 1 yangh yangh    4681 Feb 16 01:27 setup.py
-rw-rw-r-- 1 yangh yangh    3642 Feb 16 01:28 setup.pyc

**Comment by aragilar** [comment]
@Huan-Yang Those paths point to the sundials wrappers inside scikits.odes, not sundials itself (the original paths in the notebook may be wrong, it depends where benny installed sundials). You need to find where `cvode.h` was installed on your system, and the path to there.

**Comment by Huan-Yang** [comment]
@aragilar Thanks. It works for me now.

**Comment by bmcage** [comment]
Super. Closing this now. Note that we just released 2.3.2, but it is mainly documentation change release.
